src/resources/pwm_esp32.cc

```diff
@@ -119,19 +123,23 @@ class PwmResourceGroup : public ResourceGroup {
   uint32 max_value_;
 };
 
-uint32 msb(uint32 n){
-  return 31 - Utils::clz(n);
-}
-
 MODULE_IMPLEMENTATION(pwm, MODULE_PWM)
 
 PRIMITIVE(init) {
   ARGS(int, frequency, int, max_frequency)
 
-  if (frequency <= 0 || frequency > max_frequency || max_frequency > kMaxFrequency) FAIL(OUT_OF_BOUNDS);
+  uint32 src_clk_frequency = 0;
+  esp_err_t err = esp_clk_tree_src_get_freq_hz(static_cast<soc_module_clk_t>(kDefaultClk),
+                                               ESP_CLK_TREE_SRC_FREQ_PRECISION_EXACT,
+                                               &src_clk_frequency);
+  if (err != ESP_OK) return Primitive::os_error(err, process);
+
+  if (frequency <= 0 || frequency > max_frequency || max_frequency > (src_clk_frequency >> 1)) {
```

Comment:
Why do we shift `src_clk_frequency`? Add a comment?

Reply:
done.